utxo/src/cache.rs

```diff
@@ -163,18 +165,22 @@ impl<P: UtxosView> UtxosCache<P> {
         tx_undo: UtxosTxUndo,
     ) -> Result<(), Error> {
         for (i, output) in tx.outputs().iter().enumerate() {
-            let tx_outpoint = OutPoint::new(OutPointSourceId::from(tx.get_id()), i as u32);
+            let tx_outpoint = UtxoOutPoint::new(OutPointSourceId::from(tx.get_id()), i as u32);
 
             if can_be_spent(output) {
                 self.spend_utxo(&tx_outpoint)?;
             }
         }
 
         assert_eq!(tx.inputs().len(), tx_undo.inner().len());
-        for (tx_in, utxo) in tx.inputs().iter().zip(tx_undo.into_inner().into_iter()) {
-            self.add_utxo(tx_in.outpoint(), utxo, false)?;
-        }
-        Ok(())
+        tx.inputs()
+            .iter()
+            .zip(tx_undo.into_inner().into_iter())
+            .filter_map(|(tx_in, utxo)| match tx_in {
+                TxInput::Utxo(outpoint) => Some((outpoint, utxo)),
+                TxInput::Account(_) => None,
```

Comment:
I'm still not done with the review. But do we have this covered? Just keeping this here as a reminder. We need good tests for reversibility.

Reply:
nope and it doesn't work correctly, I'm on it

Reply:
fixed disconnect for utxo, added test for that